<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.5.0 -->
<title>Tutorial | DESC</title>
<meta name="generator" content="Jekyll v3.8.5" />
<meta property="og:title" content="Tutorial" />
<meta property="og:locale" content="en_US" />
<link rel="canonical" href="http://localhost:12132/desc/tutorial.html" />
<meta property="og:url" content="http://localhost:12132/desc/tutorial.html" />
<meta property="og:site_name" content="DESC" />
<script type="application/ld+json">
{"@type":"WebPage","url":"http://localhost:12132/desc/tutorial.html","headline":"Tutorial","@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/desc/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:12132/desc/feed.xml" title="DESC" /></head>
<!-- 自己插入的-->
<head>
  <script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    inlineMath: [['$','$'], ['\\(','\\)']],
    processEscapes: true
  }
});
</script>
<!-- <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"></script> -->
<script src="http://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.1/highlight.min.js"></script><script>hljs.initHighlightingOnLoad();</script><script type="text/javascript">window.MathJax = {"showProcessingMessages" : false,"messageStyle" : "none", "tex2jax": {"inlineMath": [ ["$","$"] ]}};</script><script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


</head>
  
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/desc/">DESC</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/desc/">Home</a><a class="page-link" href="/desc/installation.html">Installation</a><a class="page-link" href="/desc/tutorial.html">Tutorial</a><a class="page-link" href="/desc/reproduce.html">Reproducibility</a><a class="page-link" href="/desc/contact.html">Contact</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <p>In this tutorial, we will perform an entire <code class="highlighter-rouge">desc</code> analysis using a dataset of Peripheral Blood Mononuclear Cells (PBMC). In the dataset, the expression levels of 2,700 cells were sequenced using the Illumina NextSeq 500. The data are freely available from 10X Genomics and the raw data can be downloaded <a href="https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">here</a>.</p>

<p>Because of the difference between <code class="highlighter-rouge">tensorflow 1*</code> and <code class="highlighter-rouge">tensorflow 2*</code>, we updated our <a href="https://github.com/eleozzr/desc">desc</a> algorithm into two version such that it can be compatible with <code class="highlighter-rouge">tensorflow 1*</code> and <code class="highlighter-rouge">tensorflow 2*</code>, respectively.</p>
<ol>
  <li>For  <code class="highlighter-rouge">tensorflow 1*</code>, we released <code class="highlighter-rouge">desc(2.0.3)</code>. Please see our jupyter notebook example <a href="./jupyter_notebook/desc_2.0.3_paul.ipynb">desc_2.0.3_paul.ipynb</a></li>
  <li>For  <code class="highlighter-rouge">tensorflow 2*</code>, we released <code class="highlighter-rouge">desc(2.1.1)</code>. Please see our jupyter notebook example <a href="./jupyter_notebook/desc_2.1.1_paul.ipynb">desc_2.1.1_paul.ipynb</a></li>
</ol>

<p><br /></p>

<h1 id="0-import-python-modules">0 Import python modules</h1>

<p>We need to import the <code class="highlighter-rouge">desc</code> module for clustering analysis and <code class="highlighter-rouge">scanpy</code> for data preprocessing.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">desc</span> <span class="k">as</span> <span class="n">DESC</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="kn">import</span> <span class="nn">scanpy.api</span> <span class="k">as</span> <span class="n">sc</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="o">%</span><span class="n">matplotlib</span> <span class="n">inline</span> 
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">3</span>             <span class="c1"># verbosity: errors (0), warnings (1), info (2), hints (3)
</span><span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_versions</span><span class="p">()</span>
</code></pre></div></div>

<h1 id="1-import-data">1 Import data</h1>

<p>The current version of desc works with an <a href="https://anndata.readthedocs.io/en/latest/anndata.AnnData.html">AnnData</a> object. AnnData stores a data matrix <code class="highlighter-rouge">.X</code> together with annotations of observations <code class="highlighter-rouge">.obs</code>, variables <code class="highlighter-rouge">.var</code> and unstructured annotations <code class="highlighter-rouge">.uns</code>. The <strong>desc</strong> package provides 3 ways to prepare an AnnData object for the following analysis.</p>

<h2 id="11-start-from-a-10x-dataset">1.1 Start from a 10X dataset</h2>

<p>The desc package provides a function to load the 10X dataset directly. Download the <a href="https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">data</a> and unzip it. Then put everything in <em>filtered_gene_bc_matrices/hg19/</em> to <em>data/pbmc/</em>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load the 10X data by providing the path of the data
</span><span class="n">adata</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">read_10X</span><span class="p">(</span><span class="n">data_path</span><span class="o">=</span><span class="s">'data/pbmc'</span><span class="p">)</span>
</code></pre></div></div>

<h2 id="12-start-from-mtx-and-tsv-files">1.2 Start from <code class="highlighter-rouge">*.mtx</code> and <code class="highlighter-rouge">*.tsv</code> files</h2>

<p>When the expression data do not follow the standard 10X dataset format, we can manually import the data as follows.</p>

<p>1). Read the expression matrix from <code class="highlighter-rouge">*.mtx</code> file.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">utilities</span><span class="o">.</span><span class="n">read_mtx</span><span class="p">(</span><span class="s">'data/pbmc/matrix.mtx'</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
</code></pre></div></div>

<p>2). Read the <code class="highlighter-rouge">*.tsv</code> file for gene annotations. Make sure the gene names are unique.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'data/pbmc/genes.tsv'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s">'gene_ids'</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s">'gene_symbols'</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s">'gene_symbols'</span><span class="p">]</span>
<span class="c1"># Make sure the gene names are unique
</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">(</span><span class="n">join</span><span class="o">=</span><span class="s">"-"</span><span class="p">)</span>
</code></pre></div></div>

<p>3). Read the <code class="highlighter-rouge">*.tsv</code> file for cell annotations. Make sure the cell names are unique.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">cells</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'data/pbmc/barcodes.tsv'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'barcode'</span><span class="p">]</span> <span class="o">=</span> <span class="n">cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="c1"># Make sure the cell names are unique
</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names_make_unique</span><span class="p">(</span><span class="n">join</span><span class="o">=</span><span class="s">"-"</span><span class="p">)</span>
</code></pre></div></div>

<h2 id="13-start-from-a-h5ad-file">1.3 Start from a <code class="highlighter-rouge">*.h5ad</code> file</h2>
<p>We can also load a previously saved  <code class="highlighter-rouge">AnnData</code> object:</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s">'data/pbmc.h5ad'</span><span class="p">)</span>
<span class="c1">#or use 
#adata=sc.read_h5ad("data/pbmc.h5ad")
</span></code></pre></div></div>

<p>For more ways to create AnnData objects, please check <a href="https://anndata.readthedocs.io/en/latest/api.html#reading">here</a></p>

<p><br /></p>

<h1 id="2-preprocessing">2 Preprocessing</h1>

<p>The standard scRNA-seq data preprocessing workflow  includes filtering of cells/genes, normalization, scaling and selection of highly variables genes. In this tutorial, we use <code class="highlighter-rouge">scanpy</code> to preprocess the data. Note that among the preprocessing steps, filtration of cells/genes and selecting highly variable genes are optional, but  <strong>normalization and scaling are strictly required before the desc analysis</strong>.</p>

<p>The following steps show a typical preprocessing procedure for analyzing the PBMC data.</p>

<h2 id="21-filtering-cells-and-genes">2.1 Filtering cells and genes</h2>

<p>Typing <code class="highlighter-rouge">adata</code> in the python console shows that the original dataset contains expression values for 2700 cells across 32738 genes.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;&gt;&gt; adata
AnnData object with n_obs × n_vars = 2700 × 32738 
    obs: 'barcode', 'n_genes'
    var: 'gene_ids', 'gene_symbols', 'n_cells'
</code></pre></div></div>

<p>Let us remove cells with very few expressed genes, as well as genes are lowly expressed, using the function provided by <code class="highlighter-rouge">scanpy</code>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</code></pre></div></div>
<p>It filters out 19024 genes that are detected in less than 3 cells. There are <code class="highlighter-rouge">2700</code> cells × <code class="highlighter-rouge">13714</code> genes left.</p>

<p>A relatively high expression of mitochondrial genes indicates a lower data quality in a scRNA-seq data. Therefore, we also want to  remove cells with a high proportion of mitochondria genes expression.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mito_genes</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="nb">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">'MT-'</span><span class="p">)</span>
<span class="c1"># for each cell compute fraction of counts in mito genes vs. all genes
# the `.A1` is only necessary as X is sparse (to transform to a dense array after summing)
</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'percent_mito'</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="nb">sum</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">[:,</span> <span class="n">mito_genes</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="nb">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span>
<span class="c1"># add the total counts per cell as observations-annotation to adata
</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'n_counts'</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="nb">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span>
</code></pre></div></div>

<p>The following plots show the distribution of the number of expressed genes, the total number of UMIs and the percentage of the mitochondrial genes in each cell.</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="s">'n_genes'</span><span class="p">,</span> <span class="s">'n_counts'</span><span class="p">,</span> <span class="s">'percent_mito'</span><span class="p">],</span><span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</code></pre></div></div>
<p><img src="assets/images/output_16_0.png" alt="desc qc" /></p>

<p>Remove low quality cells according to above plots.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'n_genes'</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">2500</span><span class="p">,</span> <span class="p">:]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'percent_mito'</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">,</span> <span class="p">:]</span>
</code></pre></div></div>

<p>After filtering, we have 2638 cells and 13714 genes left in the dataset.</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;&gt;&gt; adata
AnnData object with n_obs × n_vars = 2638 × 13714 
    obs: 'barcode', 'n_genes'
    var: 'gene_ids', 'gene_symbols', 'n_cells'
</code></pre></div></div>

<h2 id="22-normalization">2.2 Normalization</h2>
<p>The function <code class="highlighter-rouge">normalize_per_cell</code> normalizes each cell by total counts of all genes, so that every cell has the same total count after normalization (10,000 by default).</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">desc</span><span class="o">.</span><span class="n">normalize_per_cell</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">counts_per_cell_after</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="c1">#or use
#sc.pp.normalize_per_cell(adata,counts_per_cell_after=1e4)
</span></code></pre></div></div>

<h2 id="23-logarithm-transformation">2.3 Logarithm transformation</h2>
<p>Also, the following analysis and the desc analysis should be performed on the log-scaled data.</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">desc</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="c1">#or use
#sc.pp.log1p(adata)
</span></code></pre></div></div>

<p>Set the <code class="highlighter-rouge">.raw</code> attribute of AnnData object to the logarithmized raw gene expression for downstream analysis, such as  differential expression analysis, and  pseudotime analysis . This simply freezes the state of the current <code class="highlighter-rouge">AnnData</code> object.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span><span class="o">.</span><span class="n">raw</span><span class="o">=</span><span class="n">adata</span>
</code></pre></div></div>

<h2 id="24-selection-of-highly-variable-genes">2.4 Selection of highly variable genes</h2>
<p>We recommend performing desc analysis on highly variable genes, which  can be selected using <code class="highlighter-rouge">highly_variable_genes</code>  function. It takes <strong>normalized, log-scaled</strong> data as input and can provide an <code class="highlighter-rouge">AnnData</code> object which contains a subset of highly variable genes.</p>

<p>The function is from <a href="https://scanpy.readthedocs.io/en/latest/">scanpy</a>. Check the <a href="https://icb-scanpy.readthedocs-hosted.com/en/latest/api/scanpy.api.pp.highly_variable_genes.html">function document</a> for detailed information about the usage and the parameter setting.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_mean</span><span class="o">=</span><span class="mf">0.0125</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code> --&gt; added
	'highly_variable', boolean vector (adata.var) 
	'means', boolean vector (adata.var)  
	'dispersions', boolean vector (adata.var)  
	'dispersions_norm', boolean vector (adata.var)  
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s">'highly_variable'</span><span class="p">]]</span>
</code></pre></div></div>

<p><strong>Note</strong>:The function <a href="https://icb-scanpy.readthedocs-hosted.com/en/latest/api/scanpy.api.pp.highly_variable_genes.html">sc.pp.highly_variable_genes</a> is slightly different from <a href="https://icb-scanpy.readthedocs-hosted.com/en/latest/api/scanpy.api.pp.filter_genes_dispersion.html#scanpy.api.pp.filter_genes_dispersion">sc.pp.filter_genes_dispersion</a>. The function <a href="https://icb-scanpy.readthedocs-hosted.com/en/latest/api/scanpy.api.pp.highly_variable_genes.html">sc.pp.highly_variable_genes</a> is similar to <code class="highlighter-rouge">FindVariableGenes</code> in R package <a href="https://satijalab.org/seurat/">Seurat</a> and it only adds some information to <code class="highlighter-rouge">adata.var</code>, but cannot filter an <code class="highlighter-rouge">AnnData</code> object automatically. Thus, if using the function <a href="https://icb-scanpy.readthedocs-hosted.com/en/latest/api/scanpy.api.pp.filter_genes_dispersion.html#scanpy.api.pp.filter_genes_dispersion">sc.pp.filter_genes_dispersion</a> , you must make sure using it after  <code class="highlighter-rouge">sc.pp.filter_genes_dispersion</code>  but before <code class="highlighter-rouge">sc.pp.log1p</code>.</p>

<p>In this dataset, 1838 genes are kept as highly variable genes.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;&gt;&gt; adata
AnnData object with n_obs × n_vars = 2638 × 1838 
    obs: 'barcode', 'n_genes', 'percent_mito', 'n_counts'
    var: 'gene_ids', 'gene_symbols', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'
</code></pre></div></div>

<h2 id="24-scaling">2.4 Scaling</h2>

<p>Finally, the data need to be standardized  before the desc analysis.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">desc</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">zero_center</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="c1">#or use
#sc.pp.scale(adata, zero_center=True, max_value=3)
</span></code></pre></div></div>

<p><br /></p>

<h1 id="3-desc-analysis">3 Desc analysis</h1>

<p>With the above data preprocessing, we are ready to run a desc analysis. The function train will perform desc on the expression matrix (2638 × 1838 in this example) and save the clustering labels as well as other related results in the <code class="highlighter-rouge">AnnData</code> object. For a full list of desc parameters please check the desc documentation on the <a href="https://pypi.org/project/desc/">pypi</a>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">dims</span><span class="o">=</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="mi">32</span><span class="p">,</span> <span class="mi">16</span><span class="p">],</span> <span class="n">tol</span><span class="o">=</span><span class="mf">0.005</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                   <span class="n">batch_size</span><span class="o">=</span><span class="mi">256</span><span class="p">,</span> <span class="n">louvain_resolution</span><span class="o">=</span><span class="p">[</span><span class="mf">0.8</span><span class="p">],</span>
                   <span class="n">save_dir</span><span class="o">=</span><span class="s">"result_pbmc3k"</span><span class="p">,</span> <span class="n">do_tsne</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">learning_rate</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span>
                   <span class="n">do_umap</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">num_Cores_tsne</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
                   <span class="n">save_encoder_weights</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</code></pre></div></div>

<p>After training of the <code class="highlighter-rouge">desc</code> model, several result slots will be added to the adata.</p>

<p>The following results are based on <code class="highlighter-rouge">desc</code> under  <strong>resolution=0.8</strong>.</p>

<ul>
  <li>The clustering result <strong>desc_0.8</strong> has been added to  <code class="highlighter-rouge">adata.obs</code>, which include clustering labels from desc.</li>
  <li>The tSNE coordinate <strong>X_tsne0.8</strong> has been added to  <code class="highlighter-rouge">adata.obsm</code>.</li>
  <li>The Umap coordinate <strong>X_umap0.8</strong> has been added to  <code class="highlighter-rouge">adata.obsm</code>.</li>
  <li>The maximum probability matrix <strong>prob_matrix0.8</strong> has been added to <code class="highlighter-rouge">adata.uns</code>.</li>
  <li><strong>X_Embeded_z0.8</strong> is the low dimensional representation of the expression data, the size is the number of  cells by the number of network nodes in the bottleneck layer (2638 x 16 in this tutorial). This slot facilitates users to generate <code class="highlighter-rouge">tSNE</code> or <code class="highlighter-rouge">Umap</code> plot or do other downstream analysis.</li>
</ul>

<p><br /></p>

<h1 id="4-visualization">4 Visualization</h1>

<p>Here, we want to check the clustering results, max probabilities, and feature plots based on the <code class="highlighter-rouge">tSNE</code> and <code class="highlighter-rouge">Umap</code>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">prob_08</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s">"prob_matrix0.8"</span><span class="p">]</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">"max.prob0.8"</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="nb">max</span><span class="p">(</span><span class="n">prob_08</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="c1">#tSNE plot 
</span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">basis</span><span class="o">=</span><span class="s">"tsne0.8"</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s">'desc_0.8'</span><span class="p">,</span><span class="s">"max.prob0.8"</span><span class="p">,</span><span class="s">'MS4A1'</span><span class="p">,</span> <span class="s">'NKG7'</span><span class="p">])</span>
</code></pre></div></div>
<p><img src="assets/images/output_45_0.png" alt="desc v1" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#Umap plot 
</span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">basis</span><span class="o">=</span><span class="s">"umap0.8"</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s">'desc_0.8'</span><span class="p">,</span><span class="s">"max.prob0.8"</span><span class="p">,</span><span class="s">'MS4A1'</span><span class="p">,</span> <span class="s">'NKG7'</span><span class="p">])</span>
</code></pre></div></div>
<p><img src="assets/images/output_46_0.png" alt="desc v2" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">"desc_0.8"</span><span class="p">]</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">"desc_0.8"</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">stacked_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="s">"MS4A1"</span><span class="p">,</span> <span class="s">"GNLY"</span><span class="p">,</span> <span class="s">"CD3E"</span><span class="p">,</span> <span class="s">"CD14"</span><span class="p">,</span> <span class="s">"FCER1A"</span><span class="p">,</span> <span class="s">"FCGR3A"</span><span class="p">,</span> <span class="s">"LYZ"</span><span class="p">,</span> <span class="s">"PPBP"</span><span class="p">,</span> <span class="s">"CD8A"</span><span class="p">],</span>
                     <span class="n">groupby</span><span class="o">=</span><span class="s">'desc_0.8'</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span><span class="n">swap_axes</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</code></pre></div></div>
<p><img src="assets/images/output_47_1.png" alt="desc v3" height="350px" />.</p>

<p><strong>Note</strong>: If you want to how to extract some information from an <code class="highlighter-rouge">AnnData</code> in R, Please <a href="./anndata_to_R.html">click here</a></p>

<p><br /></p>

<h1 id="5-save-the-result">5 Save the result</h1>
<p>Save the desc results for future analysis.</p>

<h2 id="51-save-to-a-h5ad-file">5.1 Save to a <code class="highlighter-rouge">*.h5ad</code> file</h2>
<p>The AnnData object can be saved to a <code class="highlighter-rouge">*.h5ad</code> file for the further analysis in a Python environment.</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">'../result/desc_result.h5ad'</span><span class="p">)</span>
</code></pre></div></div>

<h2 id="52-save-to-csv-files">5.2 Save to <code class="highlighter-rouge">*.csv</code> files</h2>
<p>Also, the desc result can be separatly saved to <code class="highlighter-rouge">*.csv</code> files, which can be easily accessed using <code class="highlighter-rouge">R</code>, <code class="highlighter-rouge">C</code> or other tools for the future analysis.</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#`obs` slot
</span><span class="n">meta_data</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">meta_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s">"meta.data.csv"</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">","</span><span class="p">)</span>
<span class="c1">#`obsm` slot, which is numpy.ndarray
</span><span class="n">obsm_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s">"X_tsne0.6"</span><span class="p">])</span>
<span class="n">obsm_data</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s">"tsne.csv"</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">","</span><span class="p">)</span>
</code></pre></div></div>


      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/desc/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">DESC</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">DESC</li><li><a class="u-email" href="mailto:eleozzr@gmail.com">eleozzr@gmail.com</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/eleozzr"><svg class="svg-icon"><use xlink:href="/desc/assets/minima-social-icons.svg#github"></use></svg> <span class="username">eleozzr</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p></p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
